data/transactions/verify/txn.go

```diff
@@ -362,19 +317,21 @@ func PaysetGroups(ctx context.Context, payset [][]transactions.SignedTxn, blk bo
 		case worksets <- struct{}{}:
 			if len(nextWorkset) > 0 {
 				err := verificationPool.EnqueueBacklog(ctx, func(arg interface{}) interface{} {
+					var grpErr error
 					// check if we've canceled the request while this was in the queue.
 					if tasksCtx.Err() != nil {
 						return tasksCtx.Err()
 					}
 					txnGroups := arg.([][]transactions.SignedTxn)
-					for _, signTxnsGrp := range txnGroups {
-						ctxs := PrepareContexts(signTxnsGrp, blk.BlockHeader)
-						for k, signTxn := range signTxnsGrp {
-							if err := Txn(&signTxn, ctxs[k]); err != nil {
-								return err
-							}
+					groupCtxs := make([]*GroupContext, len(txnGroups))
+					for i, signTxnsGrp := range txnGroups {
+						groupCtxs[i], grpErr = TxnGroup(signTxnsGrp, blkHeader, nil)
```

Comment:
why is the cache param nil here? `AddPayset` is used only here, so... maybe let `TxnGroup` add a group?
I do not think few additional locks make a difference there

Reply:
many transaction groups will be of size 1. I think that we shouldn't take the lock if we don't have to..
after all, taking the lock takes 3000-5000 ns; multiply this by 10000 and you'll end up with some notable delay.